File: JINSEOP/프로그래머스_혼자_풀기/숫자카드 나누기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

int get_gcd(vector<int> arr) {
    int min_v = *min_element(arr.begin(), arr.end());
    bool is_divide = true;

    while (min_v != 1) {
        is_divide = true;
        
        for (int i = 0; i < arr.size(); i++)
            if ((arr[i] % min_v) != 0) {
                is_divide = false;
                min_v -= 1;
                break;
            }

        if (is_divide)
            return min_v;
    }

    return 0;
}

bool validate(int key, vector<int> arr) {
    for (int i = 0; i < arr.size(); i++) {
        if (arr[i] % key == 0)
            return false;
    }

    return true;
}

int solution(vector<int> arrayA, vector<int> arrayB) {
    int answer = 0;
    int a_gcd = 0, b_gcd = 0;

    a_gcd = get_gcd(arrayA);
    if (a_gcd != 0) {
        if (validate(a_gcd, arrayB))
            answer = a_gcd;
    }

    b_gcd = get_gcd(arrayB);
    if (b_gcd != 0) {
        if (validate(b_gcd, arrayA))
            answer = b_gcd > a_gcd ? b_gcd : a_gcd;
    }

    return answer;
}

File: DONGUK/2023/숫자 카드 나누기.py
def gcd(a, b):
    if a % b == 0:
        return b
    return gcd(b, a % b)


def solution(arrayA, arrayB):
    answer = 0
    
    a_gcd = arrayA[0]
    N = len(arrayA)
    for i in range(1, N):
        a_gcd = gcd(a_gcd, arrayA[i])
        
    b_gcd = arrayB[0]
    for i in range(1, N):
        b_gcd = gcd(b_gcd, arrayB[i])
    
    if a_gcd > 1:
        a_flag = True
        for num in arrayB:
            if num % a_gcd == 0:
                a_flag = False
                break

        if a_flag:
            answer = a_gcd
    
    if b_gcd > 1:
        b_flag = True
        for num in arrayA:
            if num % b_gcd == 0:
                b_flag = False
                break

        if b_flag:
            answer = max(answer, b_gcd)
    return answer